protocol/messenger_communities.go

```diff
@@ -4205,3 +4216,88 @@ func chunkAttachmentsByByteSize(slice []*protobuf.DiscordMessageAttachment, maxF
 	}
 	return chunks
 }
+
+// GetCurrentKeyForGroup returns the latest key timestampID belonging to a key group
+func (m *Messenger) GetCurrentKeyForGroup(groupID []byte) (uint32, error) {
+	return m.sender.GetCurrentKeyForGroup(groupID)
+}
+
+// startCommunityRekeyLoop creates a 5-minute ticker and starts a routine that attempts to rekey every community every tick
+func (m *Messenger) startCommunityRekeyLoop() {
+	logger := m.logger.Named("CommunityRekeyLoop")
+	// TODO remove this after resolving key_id issue
+	if m.getCurrentTimeInMillis() > 0 { // Always return
+		return
+	}
+
+	var d time.Duration
+	if m.communitiesManager.RekeyInterval != 0 {
+		d = m.communitiesManager.RekeyInterval / 10
+	} else {
+		d = 5 * time.Minute
+	}
+
+	ticker := time.NewTicker(d)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				logger.Debug("rekeyAllCommunities ticker fired")
+				m.rekeyAllCommunities(logger)
+				logger.Debug("rekeyAllCommunities ticker loop ended")
+			case <-m.quit:
+				ticker.Stop()
+				logger.Debug("CommunityRekeyLoop stopped")
+				return
+			}
+		}
+	}()
+}
+
+// rekeyAllCommunities attempts to rekey every community in persistence.
+// A community will be rekeyed if it meets all the following criteria:
+//   - Community.IsAdmin()
+//   - Community.Encrypted()
+//   - Community.RekeyedAt().Add(rki).Before(time.Now()) where rki is a defined rekey interval
+func (m *Messenger) rekeyAllCommunities(logger *zap.Logger) {
+	// Determine the rekey interval, if the value is not set as a property of m.communitiesManager
+	// default to one hour
+	// TODO in future have a community level rki rather than a global rki
+	var rki time.Duration
+	if m.communitiesManager.RekeyInterval == 0 {
+		rki = time.Hour
+	} else {
+		rki = m.communitiesManager.RekeyInterval
+	}
+
+	// Get and loop over all communities in persistence
+	cs, err := m.Communities()
+	if err != nil {
+		logger.Error("error getting communities", zap.Error(err))
+		return
+	}
+	for _, c := range cs {
+		keyTimestampID, err := m.GetCurrentKeyForGroup(c.ID())
+		if err != nil {
+			logger.Error("error getting current keyTimestampID for community", zap.Error(err), zap.Binary("community ID", c.ID()))
+			continue
+		}
+
+		// TODO add functionality to encryptor.go that compares the timestamps and returns a bool
+		// c.RekeyedAt().Add(rki).Before(time.Now())
+		// keyTimestampID + rki < time.Now()
+		deleteThisFunction(rki, keyTimestampID)
+
+		if c.IsAdmin() && c.Encrypted() { // && c.RekeyedAt().Add(rki).Before(time.Now())
```

Comment:
```suggestion
		if c.IsControlNode() && c.Encrypted() { // && c.RekeyedAt().Add(rki).Before(time.Now())
```

Only control node can rekey.

Reply:
Ah, a new concept that I didn't have before. Ok I'll change that